rerun_cpp/src/rerun/component_batch.hpp

```diff
@@ -294,6 +295,33 @@ namespace rerun {
         }
     };
 
+    /// Adapter from std::vector<T> where T can be converted to TComponent
+    template <typename TComponent, typename T>
+    struct ComponentBatchAdapter<TComponent, std::vector<T>> {
+        ComponentBatch<TComponent> operator()(const std::vector<T>& input) {
+            std::vector<TComponent> transformed(input.size());
+
+            std::transform(input.begin(), input.end(), transformed.begin(), [](const T& datum) {
+                return TComponent(datum);
+            });
+
+            return ComponentBatch<TComponent>::take_ownership(std::move(transformed));
+        }
+
+        ComponentBatch<TComponent> operator()(std::vector<T>&& input) {
+            std::vector<TComponent> transformed(input.size());
+
+            std::transform(
+                std::make_move_iterator(input.begin()),
+                std::make_move_iterator(input.end()),
+                transformed.begin(),
+                [](T&& datum) { return TComponent(datum); }
+            );
+
+            return ComponentBatch<TComponent>::take_ownership(std::move(transformed));
+        }
+    };
```

Comment:
I think I really really want this. But nonetheless two concerns:
* I believe this makes error messages much much worse: If you now have a vector of things that can't be converted, it will end up in this specialization and fail there
   * have to try this
   * maybe we can to an std::enable_if on constructability to fix this!
* This always does a copy even when we do have binary compatibility.
   * worse still I'm not even sure if we may hit this path if T==TComponent. Again maybe an std::enable_if to be sure of that?
   * I don't know a quick and streamlined path out of the general copy here since we _want_ it for many cases. Imho 'usability > potential copy' though so maybe not that much of a concern